rust/cargo_manifest_to_build.py

```diff
@@ -0,0 +1,561 @@
+    path=remove_path_prefix(target["src_path"] or "build.rs", context["workspace_root"]),
+    build_deps=context["build_deps"],
+    features=",".join("\"%s\"" % f for f in context["features"]),
+    version=context["version"],
+)
+
+def resolve_deps(deps, ctxt):
+    return ",".join([
+        "\"%s\"" % ctxt["dep_format"].format(
+            name=d["name"].replace("-", "_"),
+            version=ctxt["resolved_deps"][d["name"]].replace(".", "_")
+        )
+        for d in deps
+        if d["name"] in ctxt["resolved_deps"]
+        ])
+
+def extend_context(ctxt, json):
+    out_dir_tar = ""
+    if any([t["kind"] == ["custom-build"]] for t in json["targets"]):
+        out_dir_tar = "\n    out_dir_tar = \":%s_build_script_executor\"," % json["name"].replace("-", "_")
+    data = ""
+    if ctxt["data"]:
+        data = "\n    data = r\"\"\"%s\"\"\"," % ctxt["data"]
+    for d in json["dependencies"]:
+        if d["name"] not in ctxt["resolved_deps"]:
+            sys.stderr.write("WARNING: Cannot resolve dependency on crate '%s'\n" % d["name"])
+    return {
+        "name": json["name"],
+        "workspace_root": ctxt["workspace_root"],
+        "flags": ctxt["flags"],
+        "version": json["version"],
+        "other": "%s%s" % (out_dir_tar, data),
+        "deps": resolve_deps([d for d in json["dependencies"] if not d["kind"]], ctxt),
+        "build_deps": resolve_deps([d for d in json["dependencies"] if d["kind"] == "build"], ctxt),
+        "features": json["features"].keys(),
+    }
+
+def to_build_file(ctxt, json):
+    result = ["""
+package(default_visibility = ["//visibility:public"])
+
+licenses([
+%s
+])
+
+load(
+    "@io_bazel_rules_rust//rust:rust.bzl",
+    "rust_library",
+    "rust_binary",
+)
+load(
+    "@io_bazel_rules_rust//rust:crate_repository.bzl",
+    "cargo_build_script_run",
+)
+""" % licenses(json)]
+    context = extend_context(ctxt, json)
+    for target in json["targets"]:
+        if target["kind"][0] in ["lib", "proc-macro", "dylib", "rlib"]:
+            result.append(rust_library(target, context))
+        elif target["kind"][0] == "bin":
+            result.append(rust_binary(target, context))
+        elif target["kind"][0] == "custom-build":
+            result.append(custom_build_script(target, context))
+        else:
+            result.append("# Unsupported target %s with type %s omitted" % (target["name"], target["kind"][0]))
+    return "\n\n".join(result)
+
+def parse_args_and_build_context(json):
+    args = parser.parse_args()
+    ctxt = {
+        "workspace_root": json["workspace_root"],
+        "data": args.data,
+        "dep_format": args.dep_format,
+        "resolved_deps": {v.split("=", 1)[0]: v.split("=", 1)[1] for v in args.dep},
+        "output": args.output,
+        "flags": args.flag or [],
+    }
+    return ctxt
+
+if __name__ == "__main__":
+    json = json.load(StringIO.StringIO(subprocess.check_output([
+        "cargo",
+        "metadata",
```

Comment:
This invocation strikes me as "fishy". You're largely using locally generated crate metadata, and expecting it to match what cargo-raze used to generate the exact set of dependencies.

In isolation, you cannot trust the metadata emitted by `cargo metadata` -- you need to run this for the entire workspace and then select the metadata for a single crate.

Reply:
Sorry I fail to follow here. Each crate declare its direct deps and we rely on external source to provide us the locked version, why would this invocation be wrong? I meant it barely exposes the information in the Cargo.toml file in a JSON file, it does not do any resolution.

Reply:
The locked crate version isn't the only piece of non-local state that needs to be brought in from external sources. You also need to bring in the set of features wanted by all upstream entities.

Consider a scenario where you have the following dependency chain:
A -(wants)-> B -> C

They use the default feature set. Now, consider a new crate D such that
A -> D -> C
... Where D adds a feature directive to it's C where ["with_serde"] is a wanted C feature.

This is what I called a "non-local" effect -- this new feature flag must also be provided to the C repository rule declaration... with the addition of the new locked serde deps.

As a final wrench to throw in -- this can happen per target if C is a ` [target.*.dependencies]` target-specific dependency of D.